### PR TITLE
Adds `onReady` hook to Spiderable package

### PR DESCRIPTION
**Pull request by RobertLowe**
Resolves #3824
Adds missing `DDP` dependancy
Adds `callback-hook` for onReady hook registration
Adds some client tests for `Spiderable`
Bumps version to `1.0.8`

### CODE REVIEW

File: packages/spiderable/spiderable_client.js [hunk 1]
```diff
@@ -17,14 +17,52 @@
 
 Spiderable._initialSubscriptionsStarted = false;
 
+Spiderable._onReadyHook = new Hook({
+  debugPrintExceptions: "onReadyHook callback"
```
Comment (glasser) [inline packages/spiderable/spiderable_client.js]: Probably should say `Spiderable.onReady callback`

File: packages/spiderable/spiderable_client.js [hunk 2]
```diff
@@ -17,14 +17,52 @@
...
+});
+
+// register a new onReady hook for validation
+Spiderable.onReady = function (fn) {
+  var self = this;
```
Comment (glasser) [inline packages/spiderable/spiderable_client.js]: I don't find the use of `self` when it's not really an object method to be clearer than just writing `Spiderable`. That said, I think `Spiderable._onReadyHook` could just be `var onReadyHook`.

File: packages/spiderable/spiderable_client.js [hunk 3]
```diff
@@ -17,14 +17,52 @@
...
+  return self._onReadyHook.register(fn);
+};
+
+//
+// register default hooks
+
+// top level code ready
+Spiderable.onReady(function(){
```
Comment (glasser) [inline packages/spiderable/spiderable_client.js]: Meteor style places a space before and after the `()` here.

File: packages/spiderable/spiderable_client.js [hunk 4]
```diff
@@ -17,14 +17,52 @@
...
 // This extra indirection is how we get called last
 var topLevelCodeDone = function () {
   // We'd like to use Meteor.startup here I think, but docs/behaviour of that is wrong
   Meteor._setImmediate(function () { startupCallbacksDone(); });
 };
-
 Meteor.startup(function () { topLevelCodeDone(); });
+
+// all ddp subs ready
+Spiderable.onReady(function(){
```
Comment (glasser) [inline packages/spiderable/spiderable_client.js]: ditto re space

File: packages/spiderable/spiderable_client.js [hunk 5]
```diff
@@ -17,14 +17,52 @@
...
+  Tracker.flush();
+  return DDP._allSubscriptionsReady();
+})
+
+// run all hooks and return true if they all pass
+Spiderable.isReady = function(){
```
Comment (glasser) [inline packages/spiderable/spiderable_client.js]: ditto re space

File: packages/spiderable/spiderable_client.js [hunk 6]
```diff
@@ -17,14 +17,52 @@
...
+  var self = this;
+  var isReady = true;
+  self._onReadyHook.each(function (callback) {
+    if (callback()){
```
Comment (glasser) [inline packages/spiderable/spiderable_client.js]: space before `{`

File: packages/spiderable/spiderable_client_tests.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+Tinytest.add("spiderable - is not ready while initial subscriptions aren't started", function (test, expect) {
+  Spiderable._initialSubscriptionsStarted = false;
+  test.equal(
+    Spiderable.isReady(),
+    false
+  );
+});
+
+Tinytest.add("spiderable - is not ready while DDP Subscriptions aren't ready", function (test, expect) {
+  original = DDP._allSubscriptionsReady;
```
Comment (glasser) [inline packages/spiderable/spiderable_client_tests.js]: missing `var`

File: packages/spiderable/spiderable_client_tests.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+  Spiderable._initialSubscriptionsStarted = true;
+  DDP._allSubscriptionsReady = function(){return false};
```
Comment (glasser) [inline packages/spiderable/spiderable_client_tests.js]: missing spaces and semicolon after `false`

File: packages/spiderable/spiderable_client_tests.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
+Tinytest.add("spiderable - default hooks registered", function (test, expect) {
+  test.equal(
+    _.keys(Spiderable._onReadyHook.callbacks).length,
+    2
+  );
+});
+
+Tinytest.add("spiderable - is not ready while initial subscriptions aren't started", function (test, expect) {
+  Spiderable._initialSubscriptionsStarted = false;
+  test.equal(
```
Comment (glasser) [inline packages/spiderable/spiderable_client_tests.js]: can use `test.isFalse`

File: packages/spiderable/spiderable_client_tests.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
+Tinytest.add("spiderable - default hooks registered", function (test, expect) {
+  test.equal(
+    _.keys(Spiderable._onReadyHook.callbacks).length,
+    2
+  );
+});
+
+Tinytest.add("spiderable - is not ready while initial subscriptions aren't started", function (test, expect) {
+  Spiderable._initialSubscriptionsStarted = false;
```
Comment (glasser) [inline packages/spiderable/spiderable_client_tests.js]: There's nothing that reverts this when this test is done. Tests shouldn't leave side effects in global state.

File: packages/spiderable/spiderable_client_tests.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    true
+  );
+
+  // restore original
+  DDP._allSubscriptionsReady = original;
+});
+
+Tinytest.add("spiderable - is not ready with a custom hook", function (test, expect) {
+  //clear all callbacks
+  _.each(Spiderable._onReadyHook.callbacks, function(value,key,list){
```
Comment (glasser) [inline packages/spiderable/spiderable_client_tests.js]: There's nothing that reverts this when this test is done. Tests shouldn't leave side effects in global state.

File: packages/spiderable/spiderable_client_tests.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+  });
+  test.equal(
+    Spiderable.isReady(),
+    false
+  );
+});
+
+Tinytest.add("spiderable - is ready with a custom hook", function (test, expect) {
+  //clear all callbacks
+  _.each(Spiderable._onReadyHook.callbacks, function(value,key,list){
```
Comment (glasser) [inline packages/spiderable/spiderable_client_tests.js]: There's nothing that reverts this when this test is done. Tests shouldn't leave side effects in global state.

File: packages/spiderable/spiderable_client.js [hunk 13]
```diff
@@ -17,14 +17,52 @@
 
 Spiderable._initialSubscriptionsStarted = false;
 
+Spiderable._onReadyHook = new Hook({
+  debugPrintExceptions: "onReadyHook callback"
+});
+
+// register a new onReady hook for validation
+Spiderable.onReady = function (fn) {
```
Comment (glasser) [inline packages/spiderable/spiderable_client.js]: I think this name is confusing; it sounds more like "callbacks to call when things become ready".  Maybe `Spiderable.addReadyCondition`?

**Comment by glasser** [comment]
I've added a few review comments.

File: packages/spiderable/spiderable_client.js (hunk 13 above)
Comment (RobertLowe) [inline packages/spiderable/spiderable_client.js]: I have no preferences on what the hook or registering function are named, please just provide me which names you'd like to use for each.

File: packages/spiderable/spiderable_client.js (hunk 2 above)
Comment (RobertLowe) [inline packages/spiderable/spiderable_client.js]: I have no preferences. Sure, we can just use `this` if you like?

File: packages/spiderable/spiderable_client.js (hunk 3 above)
Comment (RobertLowe) [inline packages/spiderable/spiderable_client.js]: Sure...

File: packages/spiderable/spiderable_client_tests.js (hunk 10 above)
Comment (RobertLowe) [inline packages/spiderable/spiderable_client_tests.js]: I completely agree. A root issue is that `tinytest` provides zero utility for mocking/stubbing and in addition provides zero utility for setup and teardown (which would be really *really* nice). This was more so a shot in the dark and I wanted to at least give the `spiderable` *some* tests.

File: packages/spiderable/spiderable_client_tests.js (hunk 9 above)
Comment (RobertLowe) [inline packages/spiderable/spiderable_client_tests.js]: Thanks, will do.

File: packages/spiderable/spiderable_client_tests.js (hunk 7 above)
Comment (RobertLowe) [inline packages/spiderable/spiderable_client_tests.js]: Nice catch! Thanks.

**Comment by RobertLowe** [comment]
@glasser thanks for the review! I'll go through and make the recommended changes.

**Comment by RobertLowe** [comment]
@glasser updated with recommendations

Cheers

**Comment by glasser** [comment]
Thanks, merged!

**Pull request closed without merging**